### DISCUSSION TITLE
ILLink

### DISCUSSION THREAD

**Discussion by tranb3r** [SixLabors/ImageSharp#2640] [Q&A] [answered]
ImageSharp is one of the biggest assemblies in my android app.
I've enabled dotnet trimmer (ILLink), but it only reduces the ImageSharp assembly size by ~30%.
I really have a basic usage (load an image, perform a simple computation on it), so I'm surprised the linker does not remove more code from ImageSharp.
Are you already aware of this issue?
Is there anything I can do to improve the efficency of the linker (for example, if I want to support only Png and drop all the other formats)?

**Comment by JimBobSquarePants** [comment]
Yes. Are you using the latest version? We recently did work that much reduced the binary size. https://github.com/SixLabors/ImageSharp/pull/2514

**Reply by tranb3r** [reply]
Yes, I'm using the latest version.

I've tested several things but I can't figure out how to do it properly.

```
var img = Image.Load<Rgba32>(stream);
var img = Image.Load<Rgba32>(new DecoderOptions { Configuration = new() }, stream);
var img = Image.Load<Rgba32>(new DecoderOptions { Configuration = new Configuration(new PngConfigurationModule()) }, stream);
var img = PngDecoder.Instance.Decode<Rgba32>(new PngDecoderOptions { GeneralOptions = new DecoderOptions { Configuration = new() }}, stream);
```

Could you help me figuring out what is the proper implementation for loading a Png image from streamn (and being compliant with the trimmer)?

**Reply by tranb3r** [reply]
I've also tested a piece of code that I've seen [here](https://github.com/SixLabors/ImageSharp/pull/2514#issuecomment-1748077828), but I don't get the same results: formats (bmp, gif, jpeg...) are not trimmed from the assembly.

**Reply by tranb3r** [reply]
I've spent a few more minutes on this and here is my conclusion.

The `Configuration` class contains this code that is not removed by the trimmer:
```
public static Configuration Default { get; } = Lazy.Value;

private static readonly Lazy<Configuration> Lazy = new Lazy<Configuration>(new Func<Configuration>(Configuration.CreateDefaultInstance));

internal static Configuration CreateDefaultInstance()
{
	return new Configuration(new IImageFormatConfigurationModule[]
	{
		new PngConfigurationModule(),
		new JpegConfigurationModule(),
		new GifConfigurationModule(),
		new BmpConfigurationModule(),
		new PbmConfigurationModule(),
		new TgaConfigurationModule(),
		new TiffConfigurationModule(),
		new WebpConfigurationModule(),
		new QoiConfigurationModule()
	});
}
```

**Reply by JimBobSquarePants** [reply]
I’ll re-check my output. Can’t remove the lazy evaluation without a major version though.

**Comment by JimBobSquarePants** [comment]
If you’re not getting the same results then you’re doing something different. I built that project and ran Sizoscope against it. 

Could you share your results?

**Reply by tranb3r** [reply]
When I build the project and run Sizoscope, I get the same result as you do.
However, if I modify the csproj and use `PublishTrimmed` instead of `PublishAot`, I get a different result.
Any idea why PublishTrimmed and PublishAot give a different result in terms of code trimming?

Screenshot of assembly in dnspy after publishing with trimmer enabled:
![image](https://github.com/SixLabors/ImageSharp/assets/12516566/65420e99-c2a3-4c00-820b-32ddcde564c1)

**Reply by JimBobSquarePants** [reply]
@MichalStrehovsky would you be able to assist with an explaination here? I've read the docs for both properties and am left without an understandoing of why they produce different results.

**Reply by MichalStrehovsky** [reply]
This is a bug in ILLink (a component that does IL-level trimming for PublishTrimmed). PublishAot doesn't use ILLink since it does trimming at the level of native artifacts, not IL.

I filed it as https://github.com/dotnet/runtime/issues/96655, you can track it there.

The way I root caused it:

* We have a tool similar to sizoscope that is more general purpose (works for both AOT and ILLink): https://github.com/dotnet/runtime/tree/b4ec422386bd358602c7693ff4ab28c96a806ec0/src/coreclr/tools/aot/DependencyGraphViewer. (But it doesn't tell anything about size, just "what is in the output and why")
* To generate a log that can be used with this, add `<_ExtraTrimmerArgs>--dump-dependencies --dependencies-file-format dgml</_ExtraTrimmerArgs>` to the project file
* This spits out a DGML file into obj folder after PublishTrimmed is done
* Open the file in DependencyGraphViewer and search for `CreateDefaultInstance` (the thing that _should not be_ in a trimmed app)
* Notice it was included because Configuration::.cctor uses it (the static constructor)
* This one was included because Configuration::.ctor accesses a field. Looking at the IL, we can see it accessed a field, but not a _static field_.
* Ergo the static constructor was included erroneously, accessing a non-static field doesn't need the static constructor

If the underlying issue is fixed, the size will shrink from 700-something kB to 200-something kB.

You could technically work around by moving the static constructor to a different type and have the properties forward to underlying storage there (so that Configuration doesn't have a static constructor)... but might not be worth the hassle.

**Reply by tranb3r** [reply]
Good catch! Thanks a lot.

**Reply by tranb3r** [reply]
@MichalStrehovsky 
Thanks for the fix!
Is it going to be released with dotnet 8.0.2?

**Reply by MichalStrehovsky** [reply]
In .net 9. You'd need to comment on the issue and make a _very compelling_ case for why you need it in 8. Perf issues that are not regressions from 7.0 need to be very significant to be even considered. People above my pay grade decide these.

**Reply by tranb3r** [reply]
@JimBobSquarePants what do you think?
- is this something that could be fixed before .net9?
- does it have to be a breaking change? 
- are you planning a major version soon?

**Reply by JimBobSquarePants** [reply]
Thanks so much @MichalStrehovsky for your insight here and in the identification of the issue! You rock! 

@tranb3r I am planning a major version this year but this is likely at least 6 months away as I’ve a huge amount of work to go through. It would, unfortunately, have to be a breaking change to workaround the issue pre .NET 9.

**Reply by tranb3r** [reply]
I understand.
I'll let you decide if you'd rather fix it in the future major version or wait for .net9.
Thanks to both of you for your help!

**Reply by MichalStrehovsky** [reply]
I think you can work around it with something along the lines of:

```diff
-private static readonly Lazy<Configuration> Lazy = new(CreateDefaultInstance);
+private static Lazy<Configuration> Lazy => LazyConfiguration.Lazy;
+
+static class LazyConfiguration
+{
+    internal static readonly Lazy<Configuration> Lazy = new(CreateDefaultInstance);
+}
```

I didn't actually try this but the issue is basically that right now, accessing an instance field brings in the static constructor on this class. Moving the expensive part into a separate class (with a separate static constructor) will work around because nobody is going to access the `Lazy` property on `Configuration` and the entire `LazyConfiguration` class should get trimmed.

**Reply by akoeplinger** [reply, marked as answer]
@tranb3r another workaround that you can try in your project is to use an [ILLink substitution xml](https://github.com/dotnet/runtime/blob/main/docs/tools/illink/data-formats.md#substitution-format) to replace the static constructor body with an empty stub:

Create a file called substitution.xml
```xml
<linker>
  <assembly fullname="SixLabors.ImageSharp">
    <type fullname="SixLabors.ImageSharp.Configuration">
      <method signature="System.Void .cctor()" body="stub" />
    </type>
  </assembly>
</linker>
```

And add the following to your .csproj: 

```xml
<_ExtraTrimmerArgs>--substitutions substitution.xml</_ExtraTrimmerArgs>
```

**Reply by tranb3r** [reply]
Wow!! Thanks @akoeplinger, this is a great suggestion.
It totally works for me, without any change in ImageSharp or ILLink.

**Reply by MichalStrehovsky** [reply]
To be clear, this replaces the static constructor with an empty method. There will be no warning if at some point your use of ImageSharp needs this static constructor. Make sure you test your app well because your app would only be broken after trimming, not during F5 debugging.

**Reply by tranb3r** [reply]
Yes. Actually I'm already familiar with ILLink substitutions. Thanks.

**Reply by JimBobSquarePants** [reply]
@tranb3r if you fancy trying the code change suggested in a fork I can add it to the next bug fix release if successful.

**Reply by tranb3r** [reply]
One more question for the ILLink experts. Nothing alarming though.

I have created both a maui app for android and a console app, with the same ImageSharp code (load a png, save a bmp).
But when I trim these apps (with PublishTrimmed and the ILLink substitutions hack) I do not get exactly the same result for ImageSharp.dll:
- in the maui app, several methods are not removed by the linker, but there content is replaced with `throw new NotSupportedException ("Linked Away");`
- whereas in the console app, these methods are removed by the linker.

Any explanation for this behavior? 
I'm pretty sure ImageSharp has no platform specific code, right?

**Reply by MichalStrehovsky** [reply]
You'd typically see the Linked away bodies if the type e.g. implements an interface, the interface method is called (and cannot be removed from the interface itself), but the type implementing the interface was not `new`d. Removing the method completely would fail loading the type (because it doesn't implement the interface it claims to implement), but the interface method can be implemented with a stubbed out method. The console app probably doesn't call the interface method and it can be removed from the interface.

**Reply by tranb3r** [reply]
Ok but the console app and the android app call the same code (it's in a library). So I would expect the same trimming to happen. Which does not seem to be the case.
I can share a repro if you want to investigate.

**Comment by tranb3r** [comment]
> @tranb3r if you fancy trying the code change suggested in a fork I can add it to the next bug fix release if successful.

The suggested code change does not work on my repro app. 
I must be missing something...

**Reply by MichalStrehovsky** [reply]
If you have the ImageSharp library with the change, I can take a look.

**Reply by tranb3r** [reply]
You can review the change I've tested here: https://github.com/tranb3r/ImageSharp/commit/5f59edf08efa7768cfd2cfae9c6de45cf84cedfc
What looks wrong to me is that we still have this code in the `Configuration` class:
`public static Configuration Default { get; } = Lazy.Value;`
